src/data/nat/lattice.lean

```diff
@@ -0,0 +1,153 @@
+
+lemma eq_Ici_of_nonempty_of_upward_closed {s : set ℕ} (hs : s.nonempty)
+  (hs' : ∀ (k₁ k₂ : ℕ), k₁ ≤ k₂ → k₁ ∈ s → k₂ ∈ s) : s = Ici (Inf s) :=
+ext (λ n, ⟨λ H, nat.Inf_le H, λ H, hs' (Inf s) n H (Inf_mem hs)⟩)
+
+lemma Inf_upward_closed_eq_succ_iff {s : set ℕ}
+  (hs : ∀ (k₁ k₂ : ℕ), k₁ ≤ k₂ → k₁ ∈ s → k₂ ∈ s) (k : ℕ) :
+  Inf s = k + 1 ↔ k + 1 ∈ s ∧ k ∉ s :=
+begin
+  split,
+  { intro H,
+    rw [eq_Ici_of_nonempty_of_upward_closed (nonempty_of_Inf_eq_succ H) hs, H, mem_Ici, mem_Ici],
+    exact ⟨le_refl _, k.not_succ_le_self⟩, },
+  { rintro ⟨H, H'⟩,
+    rw [Inf_def (⟨_, H⟩ : s.nonempty), find_eq_iff],
+    exact ⟨H, λ n hnk hns, H' $ hs n k (lt_succ_iff.mp hnk) hns⟩, },
+end
+
+/-- This instance is necessary, otherwise the lattice operations would be derived via
+conditionally_complete_linear_order_bot and marked as noncomputable. -/
+instance : lattice ℕ := lattice_of_linear_order
+
+noncomputable instance : conditionally_complete_linear_order_bot ℕ :=
+{ Sup := Sup, Inf := Inf,
+  le_cSup    := assume s a hb ha, by rw [Sup_def hb]; revert a ha; exact @nat.find_spec _ _ hb,
+  cSup_le    := assume s a hs ha, by rw [Sup_def ⟨a, ha⟩]; exact nat.find_min' _ ha,
+  le_cInf    := assume s a hs hb,
+    by rw [Inf_def hs]; exact hb (@nat.find_spec (λn, n ∈ s) _ _),
+  cInf_le    := assume s a hb ha, by rw [Inf_def ⟨a, ha⟩]; exact nat.find_min' _ ha,
+  cSup_empty :=
+  begin
+    simp only [Sup_def, set.mem_empty_eq, forall_const, forall_prop_of_false, not_false_iff,
+      exists_const],
+    apply bot_unique (nat.find_min' _ _),
+    trivial
+  end,
+  .. (infer_instance : order_bot ℕ), .. (lattice_of_linear_order : lattice ℕ),
+  .. (infer_instance : linear_order ℕ) }
+
+section
+
+variables {α : Type*} [complete_lattice α]
+
+lemma supr_lt_succ (u : ℕ → α) (n : ℕ) : (⨆ k < n + 1, u k) = (⨆ k < n, u k) ⊔ u n :=
+by simp [nat.lt_succ_iff_lt_or_eq, supr_or, supr_sup_eq]
+
+lemma supr_lt_succ' (u : ℕ → α) (n : ℕ) : (⨆ k < n + 1, u k) = u 0 ⊔ (⨆ k < n, u (k + 1)) :=
+by { rw ← sup_supr_nat_succ, simp }
+
+lemma infi_lt_succ (u : ℕ → α) (n : ℕ) : (⨅ k < n + 1, u k) = (⨅ k < n, u k) ⊓ u n :=
+@supr_lt_succ (order_dual α) _ _ _
+
+lemma infi_lt_succ' (u : ℕ → α) (n : ℕ) : (⨅ k < n + 1, u k) = u 0 ⊓ (⨅ k < n, u (k + 1)) :=
+@supr_lt_succ' (order_dual α) _ _ _
+
+end
+
+end nat
+
+namespace set
+
+variable {α : Type*}
+
+lemma bUnion_lt_succ (u : ℕ → set α) (n : ℕ) : (⋃ k < n + 1, u k) = (⋃ k < n, u k) ∪ u n :=
+nat.supr_lt_succ u n
+
+lemma bUnion_lt_succ' (u : ℕ → set α) (n : ℕ) : (⋃ k < n + 1, u k) = u 0 ∪ (⋃ k < n, u (k + 1)) :=
+nat.supr_lt_succ' u n
+
+lemma bInter_lt_succ (u : ℕ → set α) (n : ℕ) : (⋂ k < n + 1, u k) = (⋂ k < n, u k) ∩ u n :=
+nat.infi_lt_succ u n
+
+lemma bInter_lt_succ' (u : ℕ → set α) (n : ℕ) : (⋂ k < n + 1, u k) = u 0 ∩ (⋂ k < n, u (k + 1)) :=
+nat.infi_lt_succ' u n
+
+end set
+
+namespace enat
+open_locale classical
+
+noncomputable instance : complete_linear_order enat :=
```

Comment:
Does it make more sense to put this in `data.nat.enat` and change the imports a bit? Or does that not work?

Reply:
I have no opinion here. Actually, I wanted to add `nat.supr_lt_succ` etc and there was no good home for these lemmas, so I moved code around. If you want me to make `data.nat.enat` import `data.nat.lattice`, then I can do it. OTOH, having all `*nat` (conditionally) complete lattice instances in one file makes sense too.

Reply:
I also don't care and don't think it matters much, just caught my eye that this was the only mention of `enat` in the new file and it came at the very end. Up to you.